transport/src/commonMain/kotlin/com/genesys/cloud/messenger/transport/core/StateMachineImpl.kt

```diff
@@ -0,0 +1,72 @@
+package com.genesys.cloud.messenger.transport.core
+
+import com.genesys.cloud.messenger.transport.core.MessagingClient.State
+import com.genesys.cloud.messenger.transport.util.logs.Log
+import com.genesys.cloud.messenger.transport.util.logs.LogTag
+
+internal class StateMachineImpl(
+    val log: Log = Log(enableLogs = false, LogTag.STATE_MACHINE),
+) : StateMachine {
+    private val canConnect: Boolean by lazy { currentState is State.Closed || currentState is State.Idle || currentState is State.Error || currentState is State.Reconnecting }
+    private val canConfigure: Boolean by lazy { currentState is State.Connected || currentState is State.Reconnecting }
+    private val isReconnecting: Boolean by lazy { currentState is State.Reconnecting }
+    private val canDisconnect: Boolean by lazy { currentState !is State.Closed && currentState !is State.Idle && currentState !is State.Error }
+
+    override var currentState: State = State.Idle
```

Comment:
I think we should set the `field` before invoking the listeners so that the value of the `currentState` property matches the StateChange's newState from the perspective of the SDK user at the time the listener is invoked.

```kotlin
override var currentState: State = State.Idle
        set(value) {
            if (field != value) {
                log.i { "State changed from: $field, to: $value" }
                val oldValue = field
                field = value
                stateListener?.invoke(value)
                onStateChanged?.invoke(StateChange(oldState = oldValue, newState = value))
            }
        }
```